### ISSUE TITLE
HttpMethod should include "PATCH" as a property

### ISSUE THREAD

**Issue by ChaseFlorell** [dotnet/corefx#17299]
# Latest Proposal

Expose Http `PATCH` Verb alongside [existing verbs on `HttpMethod`](https://apisof.net/catalog/System.Net.Http.HttpMethod) `GET`, `PUT`, `POST`, `DELETE`, etc. and on [`HttpClient` as Async methods](https://apisof.net/catalog/System.Net.Http.HttpClient)

```diff
 namespace System.Net.Http {
     public class HttpClient {
+        public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content);
+        public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content);
+        public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content, CancellationToken cancellationToken);
+        public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content, CancellationToken cancellationToken);
     }
     public class HttpMethod {
+        public static HttpMethod Patch { get; }
     }
 }
```

# Original Proposal

## Description

As a part of Http requests, PATCH is an often used way to do a partial update of an object, and is common enough to merit it's own property within the `HttpMethod` class.

## Proposed API Changes

```csharp
HttpMethod.Patch;
```

## Example

```csharp
// old
var request = new HttpRequestMessage
{
    Method = new HttpMethod("PATCH") // deviates from Get, Put, Post, Delete
};

// new
var request = new HttpRequestMessage
{
    Method = HttpMethod.Patch // matches Get, Put, Post, Delete
};
```

I'd like to submit a PR with the simple change to add PATCH as an Http Method in order to round out the class. This change in the future could also extend the `HttpClient` to include new `PatchAsync` methods.

**Comment by davidsh** [comment]
Setting to Future milestone since this is not critical to 2.0 release.

**Comment by karelz** [comment]
@davidsh this looks like new API addition - i.e. api-needs-work, not enhancement.

Is it something we need to discuss on v-team? If it the addition is reasonable and non-controversial from your point of view, let's just submit it to API review (api-ready-for-review).

**Comment by itanex** [comment]
@davidsh @karelz Reviewing this against the HTTP/1.1 specifications
[RFC 7231, section 4: Request methods](http://tools.ietf.org/html/7231#section-4) and [RFC 5789, section 2: Patch method](http://tools.ietf.org/html/5789#section-2), it would be wise to include the CONNECT verb is also missing.

**Comment by davidsh** [comment]
>it would be wise to include the CONNECT verb is also missing.

I don't see any value in adding the CONNECT verb as a convenience method to the HttpMethod class.  CONNECT is also used primarily during TLS/SSL connection thru a proxy.  And it is used internally while making the HTTP request for the other verbs.  The HttpClient doesn't support using the CONNECT verb on its own.

**Comment by ChaseFlorell** [comment]
If given the go ahead, I'd be willing to submit a PR for this.

Should having a patch method to accompany this be a separate issue/pr?

**Comment by davidsh** [comment]
>If given the go ahead, I'd be willing to submit a PR for this.

This needs API review/approval first.

In addition, adding this method will bump the Http APIs out of netstandard.  That means, it won't be available for Desktop either. So, changes will have to be made to ref, assembly version, contract version, etc.

You'll need to refer to guidelines for that:
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/adding-api-guidelines.md

**Comment by karelz** [comment]
@davidsh I take it from your comment that you are fine with the API shape (aka "first-level API review is fine"). Can you please confirm? If not sure, can you please raise it and discuss in Networking v-team sync up tomorrow? (in the worst case I can add it to agenda for next week)

As soon as Networking experts (@davidsh @CIPop @Priya91 @geoffkizer) are fine with the API shape, let's please mark it as "api-ready-for-review", so that we can review it at API review meeting the next Tuesday (or the one after).

Regarding your ref/assembly version/contract version comment - yes, the implementation won't be super-simple, but those things are doable. We have plenty of new .NET Core-only APIs in other assemblies, so we can follow their lead and see how to do things. Moreover we have experts at hand to help us with these technicalities.

@ChaseFlorell thank you for your interest in contribution. We appreciate it. Adding new APIs is not simple - both technically and in principle, because we have to assess APIs from the point of view "we will maintain them/live with them forever once we add them". That's the main reason why it is not a good idea to rush new APIs through, so please prepare yourself for slower-moving process (which is slower-moving on purpose).
I will try to get closure on the shape of the API on Networking v-team. Once that is done, you can submit your PR -- note that the order of steps is in place mainly to avoid your wasted time on things we might reject later (I think nobody likes that).
Also note that we are working hard towards landing 2.0 (see details in #17619). If you want us to help with that, please check #17619 how best to help. we could definitely use some help with issues in Networking area ;-). Thank you!

**Comment by davidsh** [comment]
>@davidsh I take it from your comment that you are fine with the API shape (aka "first-level API review is fine"). Can you please confirm? If not sure, can you please raise it and discuss in Networking v-team sync up tomorrow? (in the worst case I can add it to agenda for next week)

In terms of this particular API revision, adding this single new convenience method looks fine from an API shape perspective. We will discuss in team meeting.

Usually, it is best to group several API additions together for a library such as System.Net.Http. As noted above, there is large overhead in doing API changes in general. So, it is best to consider that overhead and group a few updates together.

For example: 
https://github.com/dotnet/corefx/issues/4382

is relatively simple in terms of technical things (i.e. add enum values to the HttpStatusCode enumeration).

**Comment by ChaseFlorell** [comment]
If grouping a few API changes is in order, I'd like to suggest adding a `PatchAsyc()` method (along with overloads) to coincide with this change.

**Comment by karelz** [comment]
@davidsh the "large overhead of adding APIs" is already behind us. Adding APIs these days is actually pretty easy - see e.g. #16034 or #16118. We may use couple of tricks from #16996 to add \*.netcoreapp.cs files, or maybe #ifdef if we're cross-compiling the sources (I don't think we do that), but that should be it.
If that's not the case, I'll be happy to get infra folks to help us and guide us.
From that perspective, no grouping needed, unless it makes logical sense.

I understand you are cautious after the incident with System.Net.Http for Desktop and I appreciate that. However, we are talking APIs which are .NET Core-only -- adding those is simple and most importantly, we are just doing what all other packages do, we are not inventing anything new here for System.Net.Http. So let's be cautious, let's run the how-to-create-PR-guidance and PR review by packaging experts, but let's not let caution stand in our ability to innovate on the platform which was designed for easy and fast innovation -- .NET Core.

**Comment by itanex** [comment]
> CONNECT is also used primarily during TLS/SSL connection thru a proxy. And it is used internally while making the HTTP request for the other verbs. The HttpClient doesn't support using the CONNECT verb on its own.

Perhaps I have misinterpreted the purpose of objects in the System.Net.Http namespace. As  I understand this namespace it contains objects specific to the operational needs of Http as a protocol. _Though that might just be my way of understanding it._

However, the documentation around the System.Net.Http.HttpMethod object itself does not declare it's use exclusively for HttpClient. So, hypothetically, if I were to have a need to manage the CONNECT verb in my code I should be using another object? Which object should I be using in that case? Why not supporting the Http protocol in its entirety not of value?

Help me understand.

**Comment by geoffkizer** [comment]
> As I understand this namespace it contains objects specific to the operational needs of Http as a protocol. 


I think that's correct.

**Comment by ChaseFlorell** [comment]
Did this issue just fall by the wayside?

**Comment by CIPop** [comment]
I don't have any objections to adding this API together with tests that prove that the underlying stacks (HttpClient, ASP.Net, HttpListener) support it.

> I'd like to suggest adding a PatchAsyc() method (along with overloads) to coincide with this change.

@ChaseFlorell your later message regarding PatchAsync is not clear. Do you want to add that as well or just the enum? (Please add all proposed APIs to the request.)

Some more information regarding the verb and its usage:

From the RFC standpoint, the PATCH verb is not part of [RFC7231](https://tools.ietf.org/html/rfc7231). It's mentioned once within the section describing [PUT](https://tools.ietf.org/html/rfc7231#section-4.3.4): 
```
   Partial content updates are possible by
   targeting a separately identified resource with state that overlaps a
   portion of the larger resource, or by using a different method that
   has been specifically defined for partial updates (for example, the
   PATCH method defined in [RFC5789]).
```

The abstract of https://tools.ietf.org/html/rfc5789 is stating that the PATCH method is used by applications _extending_ the HTTP protocol.

One interesting use-case is described in https://tools.ietf.org/html/rfc6902: 

```
   JavaScript Object Notation (JSON) [RFC4627] is a common format for
   the exchange and storage of structured data.  HTTP PATCH [RFC5789]
   extends the Hypertext Transfer Protocol (HTTP) [RFC2616] with a
   method to perform partial modifications to resources.

   JSON Patch is a format (identified by the media type "application/
   json-patch+json") for expressing a sequence of operations to apply to
   a target JSON document; it is suitable for use with the HTTP PATCH
   method.
```

**Comment by moljac** [comment]
Hi

My issue was dupe of this one, so I'll add my comment and links here.

Rationale: I'm Xamarin.Auth maintainer and developer and currently am investigating problems with Google Drive and Xamarin.Auth. Namely Xamarin.Auth has all verbs/methods that are as static properties in HttpMethod, but PATCH and of course Google needs PATCH for Drive API. Current workaround is to add `new HttpMethod("PATCH")`, but I it would be nice to have it as property to since it is part of the RFC.

The commit referenced will be tested next week with Google APIs.

moljac/corefx@457e528

Data copied from my issue:

System.Net.Http.HttpMethod.Patch method/verb (RFC5789) is a extension to HTTP 1.1 protocol and method definitions defined in (RFC2616 Section 9).

Suggestion: add API (static method HttpMethod.Patch that returns new HttpMethod("PATCH")) which is current workaround for this issue.

Pros:

*   consistent API with other HTTP methods.

Cons:

*   none, since this is new API, there should not be any breaking changes.

Code changes can be viewed in the commit referenced below (no PR issued yet).

Preliminary builds on MacOSX passed.

References:

*   https://tools.ietf.org/html/rfc5789
*   https://tools.ietf.org/html/rfc2616#section-9
*   moljac/corefx@457e528

**Comment by h0useRus** [comment]
Is any plans to introduce HttpMethod.Patch in 2.1 ?

**Comment by TylerBrinkley** [comment]
Here's a formal API proposal.

It is common for API's to support the PATCH HTTP Verb. It would be much more convenient if `HttpClient` and `HttpMethod` were updated to support these and since `HttpMethod` is a class would reduce allocations for current users who aren't caching their Patch `HttpMethod`.

With this implemented, what is currently being implemented like this
```c#
client.SendAsync(new HttpRequestMessage(new HttpMethod("PATCH"), requestUri) { Content = content }))
```
could now be implemented like this
```c#
client.PatchAsync(requestUri, content)
```

## Proposed API
```diff
 namespace System.Net.Http {
     public class HttpClient {
+        public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content);
+        public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content);
+        public Task<HttpResponseMessage> PatchAsync(string requestUri, HttpContent content,
            CancellationToken cancellationToken);
+        public Task<HttpResponseMessage> PatchAsync(Uri requestUri, HttpContent content,
            CancellationToken cancellationToken);
     }
     public class HttpMethod {
+        public static HttpMethod Patch { get; }
     }
 }
```

**Comment by karelz** [comment]
Given the general ok from @davidsh above, I think we can consider it as ready for API review.

Minor tweak, we should add just 2 methods with `CancellationToken cancellationToken = default(CancellationToken)`

Updating top post now ...

**Comment by TylerBrinkley** [comment]
I wasn't sure if we wanted to make the `CancellationToken` parameter optional since it wouldn't match the existing methods, but I'm fine with either.

**Comment by karelz** [comment]
I have seen the pattern on new APIs - the wider API review will correct it if it is wrong ;)

**Comment by davidsh** [comment]
>I have seen the pattern on new APIs - the wider API review will correct it if it is wrong ;)

We should match the current API pattern for the existing HttpClient APIs.

cc: @terrajobst

**Comment by terrajobst** [comment]
Looks good as proposed. BTW, I'm amazed it's only that many lines. Great API :-)

**Comment by TylerBrinkley** [comment]
I'd like to take this if that's good with you guys.

**Comment by karelz** [comment]
Go ahead, assigned to you.
BTW: The approved API was the one matching the current API pattern - so we will be adding 4 methods, not just 2 (with default cancellatinToken).

**Comment by TylerBrinkley** [comment]
Sounds good. Thanks.

**Comment by karelz** [comment]
FYI: The API review discussion was recorded - see https://youtu.be/ZrT0uOsqQlI?t=46 (6 min duration)
